File: 프로그래머스/unrated/154540. 무인도 여행/무인도 여행.cpp
#include <string>
#include <vector>
#include <queue>
#include <algorithm>

using namespace std;

int dy[4] = {1,-1,0,0};
int dx[4] = {0,0,1,-1};
bool visited[101][101];

queue<pair<int,int>> q;

int bfs(vector<string> maps, int i, int j){
    q.push({i,j});
    int cnt = maps[i][j] - '0';
    visited[i][j] = true;
    while(!q.empty()){
        int y = q.front().first;
        int x = q.front().second;
        q.pop();
        
        for(int w = 0; w < 4; w++){
            int ny = y + dy[w];
            int nx = x + dx[w];
            
            if(ny < 0 || nx < 0 || ny >= maps.size() || nx >= maps[0].size()) continue;
            
            if(maps[ny][nx] != 'X' && !visited[ny][nx]){
                cnt += maps[ny][nx] - '0';
                q.push({ny,nx});
                visited[ny][nx] = true;
            }
        }
    }
    return cnt;
}
vector<int> solution(vector<string> maps) {
    vector<int> answer;
    
    for(int i = 0; i < maps.size(); i++){
        for(int j = 0; j < maps[i].size(); j++){
            if(maps[i][j] != 'X' && !visited[i][j]){
                answer.push_back(bfs(maps,i,j));
            }
        }
    }
    
    if(answer.empty()) answer.push_back(-1);
    else{
        sort(answer.begin(),answer.end());
    }
    return answer;
}

File: 프로그래머스/2/154540. 무인도 여행/무인도 여행.java
import java.util.ArrayList;
import java.util.LinkedList;
import java.util.List;
import java.util.Queue;

class Solution {
    static final int[] dy = {1, -1, 0, 0};
    static final int[] dx = {0, 0, 1, -1};
    static boolean visited[][];

    static class Pos {
        int y;
        int x;

        public Pos(int y, int x) {
            this.y = y;
            this.x = x;
        }
    }

    public int[] solution(String[] maps) {
        List<Integer> answer = new ArrayList<>();
        visited = new boolean[maps.length][maps[0].length()];

        for (int i = 0; i < maps.length; i++) {
            for (int j = 0; j < maps[i].length(); j++) {
                if (maps[i].charAt(j) != 'X' && !visited[i][j])
                answer.add(bfs(maps, new Pos(i, j)));
            }
        }
        if (answer.size() == 0) {
            return new int[]{-1};
        }
        return answer.stream().mapToInt(Integer::intValue).sorted().toArray();
    }

    private int bfs(String[] maps, Pos pos) {
        Queue<Pos> q = new LinkedList<>();
        q.offer(pos);
        int cnt = maps[pos.y].charAt(pos.x) - '0';
        visited[pos.y][pos.x] = true;

        while (!q.isEmpty()) {
            Pos cur = q.poll();

            for (int d = 0; d < 4; d++) {
                int ny = cur.y + dy[d];
                int nx = cur.x + dx[d];

                if (ny < 0 || nx < 0 || ny >= maps.length || nx >= maps[0].length()) {
                    continue;
                }

                if (maps[ny].charAt(nx) != 'X' && !visited[ny][nx]) {
                    cnt += maps[ny].charAt(nx) - '0';
                    q.offer(new Pos(ny, nx));
                    visited[ny][nx] = true;
                }
            }
        }
        return cnt;
    }
}